bob-common/src/mapper.rs

```diff
@@ -116,22 +122,24 @@ impl Virtual {
     }
 
     pub fn get_support_nodes(&self, key: BobKey, count: usize) -> Vec<&Node> {
+        debug_assert!(count < self.nodes.len());
         trace!("get target nodes for given key");
         let target_nodes = self.get_target_nodes_for_key(key);
         trace!("extract indexes of target nodes");
-        let mut target_indexes = target_nodes.iter().map(Node::index);
-        let len = target_indexes.size_hint().0;
-        debug!("iterator size lower bound: {}", len);
+        let target_indexes: Vec<_> = target_nodes.iter().map(Node::index).collect();
         trace!("nodes available: {}", self.nodes.len());
+        let offset = self.support_nodes_offset.fetch_add(1, Ordering::Relaxed);
         self.nodes
             .iter()
             .filter_map(|(id, node)| {
-                if target_indexes.all(|i| &i != id) {
+                if !target_indexes.contains(id) {
                     Some(node)
                 } else {
                     None
                 }
             })
+            .cycle()
```

Comment:
Unfortunately, there are edge cases where a `cycle()` can result in a duplicate nodes in the result. 
Example: 3 nodes, 2 replicas (replicas are for vdisks and therefore they can be easily located on 3 nodes), quorum is 2. If data written and both replicas are not available, then this function will be called with `count = 2` and 2 nodes out of 3 will be filtered out (because they are target for specific key). This results in a single node to be repeated 2 times in the result

Reply:
This can be avoided with the following approach:
1. Create vector with capacity = `count`
2. Iterate and fill vector: `for node in self.nodes.iter().skip(offset % self.nodes.len()) { .. }`
3. If vector length is less than `count` and `offset % self.nodes.len() > 0` then additionally iterate from the beginning: `for node in self.nodes.iter().take(offset % self.nodes.len()) { ... }`